src/Passes/Source/ProfilePass/Profile.cpp

```diff
@@ -0,0 +1,680 @@
+                        // everything. In this case, we should revert to the order they appear in the
+                        // function
+                        auto bb = llvm::dyn_cast<llvm::BasicBlock>(br_instr->getOperand(i));
+                        if (bb != nullptr)
+                        {
+
+                            // Ensuring that we are not scheduling the same block twice
+                            if (blocks_queued.find(bb) == blocks_queued.end())
+                            {
+                                queue.push_back(bb);
+                                blocks_queued.insert(bb);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Deleting constants
+        for (auto& x : schedule_instruction_deletion)
+        {
+            x->eraseFromParent();
+        }
+
+        --depth_;
+
+        return true;
+    }
+
+    bool ProfilePass::isActive(llvm::Value* value) const
+    {
+        return active_pieces_.find(value) != active_pieces_.end();
+    }
+
+    void ProfilePass::runCopyAndExpand(llvm::Module& module, llvm::ModuleAnalysisManager&)
+    {
+        replacements_.clear();
+        // For every instruction in every block, we attempt a match
+        // and replace.
+        for (auto& function : module)
+        {
+            if (function.hasFnAttribute("EntryPoint"))
+            {
+                runOnFunction(function, [this](llvm::Value* value, DeletableInstructions& modifier) {
+                    return copyAndExpand(value, modifier);
+                });
+            }
+        }
+
+        // Dead code detection
+        for (auto& function : module)
+        {
+            if (function.hasFnAttribute("EntryPoint"))
+            {
+                // Marking function as active
+                active_pieces_.insert(&function);
+
+                // Detectecting active code
+                runOnFunction(function, [this](llvm::Value* value, DeletableInstructions& modifier) {
+                    return detectActiveCode(value, modifier);
+                });
+            }
+        }
+
+        applyReplacements();
+    }
+
+    void ProfilePass::applyReplacements()
+    {
+        // Applying all replacements
+
+        std::unordered_set<llvm::Value*> already_removed;
+        for (auto it = replacements_.rbegin(); it != replacements_.rend(); ++it)
+        {
+            auto instr1 = llvm::dyn_cast<llvm::Instruction>(it->first);
+
+            // Checking if by accident the same instruction was added
+            if (already_removed.find(instr1) != already_removed.end())
+            {
+                throw std::runtime_error("Instruction was already removed.");
+            }
+            already_removed.insert(instr1);
+
+            if (instr1 == nullptr)
+            {
+                llvm::errs() << "; WARNING: cannot deal with non-instruction replacements\n";
```

Comment:
When would this case occur? 
How does the tool deal with errors - what happens if the tool produces a warning (e.g. does it fail with a non-zero exit code), and what happens if it produces an error (is there a concept of an error or does it throw in that case)?

Reply:
I am not sure whether this will ever occur and whether this is even a relevant case to consider. The check is there as a santity check and will be refactored once I've written a proper error/warning logger.

Reply:
I agree with and like the idea of reporting it as an internal error.